File: code/worksheet_XC.m
T = 2000; maxLags = 100; a1 = 0.83;
x = randn(T, 1); y = randn(T, 1);
for t = 4:numel(y)-1
    y(t+1) = a1 * y(t) + x(t-3) + y(t+1); % AR(1)
end

figure(333); clf; subplot(3,1,1)
[xc, lags] = xcorr(x, y, maxLags); % raw cross correlation
plot(lags, xc)
ylabel('raw cross correlation'); grid on;

subplot(3,1,2); hold all;
[xc, lags] = xcorr(x, y, maxLags, 'unbiased'); % normalize with 1/(1-m-tau)
plot(lags, xc)
[xc, lags] = xcorr(x, y, maxLags, 'coeff'); % normalize to corr coefficient
plot(lags, xc)
legend('unbiased', 'coeff');

subplot(3,1,3); hold all;
[ac, lags] = xcorr(x, maxLags, 'coeff'); % compute autocorrelation function
plot(lags, ac);
[ac, lags] = xcorr(y, maxLags, 'coeff'); % autocorrelation for y
plot(lags, ac);
xlabel('lags'); legend('x', 'y'); ylabel('autocorrelation');

File: code/worksheet_XC.py
T = 2000; maxLags = 100; a1 = 0.83;
x = np.random.normal(0,1,(T))
y = np.random.normal(0,1,(T))
for t in range(3, len(y)-1):
    y[t+1] += a1 * y[t] + x[t-3]  # AR(1)

fig, ax = plt.subplots(figsize=(15,5), nrows=1, ncols=3)

xc = np.correlate(x, y, "full")[len(x)-maxLags:len(x)+maxLags+1]; 

ax[0].plot(np.arange(-maxLags, maxLags+1), xc)
ax[0].set_ylabel('raw cross correlation')
ax[0].set_xlabel('lags')